packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright © 2021-present 650 Industries, Inc. (aka Expo)
+
+#include "JavaScriptModuleObject.h"
+
+#include <folly/dynamic.h>
+#include <jsi/JSIDynamic.h>
+#include <react/jni/ReadableNativeArray.h>
+
+#include <utility>
+#include <tuple>
+#include <algorithm>
+
+namespace jni = facebook::jni;
+namespace jsi = facebook::jsi;
+namespace react = facebook::react;
+
+namespace expo {
+
+jni::local_ref<jni::HybridClass<JavaScriptModuleObject>::jhybriddata>
+JavaScriptModuleObject::initHybrid(jni::alias_ref<jhybridobject> jThis) {
+  return makeCxxInstance(jThis);
+}
+
+void JavaScriptModuleObject::registerNatives() {
+  registerHybrid({
+                   makeNativeMethod("initHybrid", JavaScriptModuleObject::initHybrid),
+                   makeNativeMethod("registerSyncFunction",
+                                    JavaScriptModuleObject::registerSyncFunction),
+                 });
+}
+
+std::shared_ptr<jsi::Object> JavaScriptModuleObject::getJSIObject(jsi::Runtime &runtime) {
+  if (jsiObject == nullptr) {
+    auto hostObject = std::make_shared<JavaScriptModuleObject::HostObject>(this);
+    jsiObject = std::make_shared<jsi::Object>(
+      jsi::Object::createFromHostObject(runtime, hostObject));
+  }
+
+  return jsiObject;
+}
+
+void JavaScriptModuleObject::registerSyncFunction(jni::alias_ref<jstring> name, jint args) {
+  auto cName = name->toStdString();
+  methodsMetadata.emplace(std::piecewise_construct,
```

Comment:
that could use some comment :D
also I think we could use c++17, so `methodsMetadata.try_emplace(cName, cName, args, false)` should be equivalent